File: ios-tests/MobileTaskManagerUITests/TaskFormScreen.swift
import XCTest

struct TaskFormScreen {
  let app: XCUIApplication

  var screen: XCUIElement { app.element(withId: TestIds.taskFormScreen) }
  var titleField: XCUIElement { app.element(withId: TestIds.taskTitleInput) }
  var descriptionField: XCUIElement { app.element(withId: TestIds.taskDescriptionInput) }
  var dueDateField: XCUIElement { app.element(withId: TestIds.taskDueDateInput) }
  var submitButton: XCUIElement { app.element(withId: TestIds.taskSubmitButton) }
  var backButton: XCUIElement { app.element(withId: TestIds.taskFormBackButton) }
  var titleError: XCUIElement { app.element(withId: TestIds.taskTitleError) }
  var dueDateError: XCUIElement { app.element(withId: TestIds.taskDueDateError) }

  @discardableResult
  func waitForScreen(timeout: TimeInterval = 30) -> Bool {
    screen.waitForExistence(timeout: timeout)
  }

  func setTitle(_ text: String) {
    titleField.replaceText(text)
  }

  func setDescription(_ text: String) {
    descriptionField.replaceText(text)
  }

  func titleValue() -> String {
    titleField.value as? String ?? ""
  }

  func descriptionValue() -> String {
    descriptionField.value as? String ?? ""
  }

  func dueDateValue() -> String {
    dueDateField.value as? String ?? ""
  }

  func selectPriority(_ value: String) {
    app.element(withId: TestIds.testIdForPriority(value)).tap()
  }

  func isPrioritySelected(_ value: String) -> Bool {
    app.element(withId: TestIds.testIdForPriority(value)).isSelected
  }

  func isTitleErrorVisible(timeout: TimeInterval = 5) -> Bool {
    guard titleError.waitForExistence(timeout: timeout) else {
      return false
    }
    return !titleError.label.isEmpty
  }

  func isDueDateErrorVisible(timeout: TimeInterval = 5) -> Bool {
    guard dueDateError.waitForExistence(timeout: timeout) else {
      return false
    }
    return !dueDateError.label.isEmpty
  }

  func selectQuickDate(_ option: String) {
    app.element(withId: TestIds.testIdForDueDateOption(option)).tap()
  }

  func submit() {
    submitButton.tap()
  }

  func tapBack() {
    backButton.tap()
  }

  func fillAndSubmit(
    title: String,
    description: String? = nil,
    priority: String,
    quickDate: String,
  ) {
    setTitle(title)
    if let description {
      setDescription(description)
    }
    selectPriority(priority)
    selectQuickDate(quickDate)
    submit()
  }
}

File: android-tests/src/androidTest/kotlin/com/mobiletaskmanager/TaskFormScreen.kt
package com.mobiletaskmanager

import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.ViewInteraction
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.action.ViewActions.closeSoftKeyboard
import androidx.test.espresso.action.ViewActions.replaceText
import androidx.test.espresso.action.ViewActions.scrollTo
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
import androidx.test.espresso.matcher.ViewMatchers.isSelected
import androidx.test.espresso.matcher.ViewMatchers.withText
import org.hamcrest.Matchers.not

object TaskFormScreen {

  fun waitForScreen(): ViewInteraction = waitForTestId(TestIds.taskFormScreen)

  fun assertDisplayed() {
    waitForScreen().check(matches(isDisplayed()))
  }

  fun setTitle(title: String) {
    onView(withTestId(TestIds.taskTitleInput))
        .perform(scrollTo(), replaceText(title), closeSoftKeyboard())
  }

  fun setDescription(description: String) {
    onView(withTestId(TestIds.taskDescriptionInput))
        .perform(scrollTo(), replaceText(description), closeSoftKeyboard())
  }

  fun selectPriority(priority: String) {
    onView(withTestId(TestIds.testIdForPriority(priority))).perform(scrollTo(), click())
  }

  fun selectQuickDate(option: String) {
    onView(withTestId(TestIds.testIdForDueDateOption(option))).perform(scrollTo(), click())
  }

  fun submit() {
    onView(withTestId(TestIds.taskSubmitButton)).perform(scrollTo(), click())
  }

  fun assertTitleValue(title: String) {
    onView(withTestId(TestIds.taskTitleInput)).check(matches(withText(title)))
  }

  fun assertDescriptionValue(description: String) {
    onView(withTestId(TestIds.taskDescriptionInput)).check(matches(withText(description)))
  }

  fun assertDueDateValue(dueDate: String) {
    waitForTestIdWithText(TestIds.taskDueDateInput, dueDate)
  }

  fun assertDueDateNotEmpty() {
    onView(withTestId(TestIds.taskDueDateInput)).check(matches(not(withText(""))))
  }

  fun assertPrioritySelected(priority: String) {
    onView(withTestId(TestIds.testIdForPriority(priority))).check(matches(isSelected()))
  }

  fun assertTitleErrorVisible(message: String? = null) {
    waitForTestId(TestIds.taskTitleError).check(matches(isDisplayed()))
    if (message != null) {
      onView(withTestId(TestIds.taskTitleError)).check(matches(withText(message)))
    }
  }

  fun assertDueDateErrorVisible(message: String? = null) {
    waitForTestId(TestIds.taskDueDateError).check(matches(isDisplayed()))
    if (message != null) {
      onView(withTestId(TestIds.taskDueDateError)).check(matches(withText(message)))
    }
  }

  fun fillAndSubmit(
      title: String,
      description: String? = null,
      priority: String,
      quickDate: String,
  ) {
    waitForScreen()
    setTitle(title)
    if (description != null) {
      setDescription(description)
    }
    selectPriority(priority)
    selectQuickDate(quickDate)
    submit()
  }
}
